spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java

```diff
@@ -0,0 +1,207 @@
+          ProcedureParameter.optional("options", STRING_MAP),
+          ProcedureParameter.optional("where", DataTypes.StringType)
+  };
+
+  // counts are not nullable since the action result is never null
+  private static final StructType OUTPUT_TYPE = new StructType(new StructField[]{
+      new StructField("rewritten_data_files_count", DataTypes.IntegerType, false, Metadata.empty()),
+      new StructField("added_data_files_count", DataTypes.IntegerType, false, Metadata.empty())
+  });
+
+  public static ProcedureBuilder builder() {
+    return new Builder<RewriteDataFilesProcedure>() {
+      @Override
+      protected RewriteDataFilesProcedure doBuild() {
+        return new RewriteDataFilesProcedure(tableCatalog());
+      }
+    };
+  }
+
+  private RewriteDataFilesProcedure(TableCatalog tableCatalog) {
+    super(tableCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    Identifier tableIdent = toIdentifier(args.getString(0), PARAMETERS[0].name());
+
+    return modifyIcebergTable(tableIdent, table -> {
+      RewriteDataFiles action = actions().rewriteDataFiles(table);
+
+      String strategy = args.isNullAt(1) ? null : args.getString(1);
+      String sortOrders = args.isNullAt(2) ? null : args.getString(2);
+      if (sortOrders != null && (strategy == null || strategy.equalsIgnoreCase("binpack"))) {
+        throw new IllegalArgumentException(
+                "sort_order is not accepted for binpack strategy. Please configure sort strategy to use sort_order");
+      }
+      SortOrder sortOrder = collectSortOrders(strategy, table, action, sortOrders);
+      action = checkAndApplyStrategy(action, strategy, sortOrder);
+
+      if (!args.isNullAt(3)) {
+        action = checkAndApplyOptions(args, action);
+      }
+
+      String where = args.isNullAt(4) ? null : args.getString(4);
+      action = checkAndApplyFilter(action, where);
+
+      RewriteDataFiles.Result result = action.execute();
+
+      return toOutputRows(result);
+    });
+  }
+
+  private RewriteDataFiles checkAndApplyFilter(RewriteDataFiles action, String where) {
+    if (where != null) {
+      ParserInterface sqlParser = spark().sessionState().sqlParser();
+      try {
+        Expression expression = sqlParser.parseExpression(where);
+        return action.filter(SparkExpressions.convert(expression));
+      } catch (ParseException e) {
+        throw new IllegalArgumentException("Cannot parse predicates in where option: " + where);
+      }
+    }
+    return action;
+  }
+
+  private RewriteDataFiles checkAndApplyOptions(InternalRow args, RewriteDataFiles action) {
+    Map<String, String> options = Maps.newHashMap();
+    args.getMap(3).foreach(DataTypes.StringType, DataTypes.StringType,
```

Comment:
Could this be putAll?

Reply:
args.getMap(3) is of spark `MapData` object. So cannot use putAll. Hence this logic. Used a same logic that is present in `SnapshotTableProcedure` and `MigrateTableProcedure`